### DISCUSSION TITLE
Fetching pocket base dynamically

### DISCUSSION THREAD

**Discussion by hrr-amboss** [pocketbase/pocketbase#4623] [Q&A]
Hey, I'm building a learning app for jazz licks with pocket base. I have a collection with all licks, a collection for users and a collection where user-specific progress data (e.g. self-rating) is stored for each lick's training session.

Now I want to list all licks including the user specific data. What would be the best approach with pocket base? I was able to create a view collection enriching the existing licks with user specific data but only for a static user. How could I make this dynamic? I tried parameters in the view option collection but it seems pocket base doesn't support this? Is there a way to add in a parameter when querying view collections?

Thanks for your help!

**Comment by ganigeorgiev** [comment]
> Is there a way to add in a parameter when querying view collections?

No. View collections are essentially plain [SQLite views](https://www.sqlite.org/lang_createview.html) and the query can't be parameterized at the moment.

> How could I make this dynamic?

It is difficult to suggest or recommend an alternative without details about your collections and how the data is actually related.

In general if all you want is to expand a relation you can check:
- https://pocketbase.io/docs/working-with-relations/#expanding-relations
- https://pocketbase.io/docs/working-with-relations/#back-relations

If the above doesn't help, please elaborate more on your collections structure and what is the desired behaviour (_you can export your collections from the Admin UI > Settings > Export collections_).

**Comment by hrr-amboss** [comment]
@ganigeorgiev Thanks for your response. I'm attaching my schema.

The setup I have
- cells: collection containing the data for all jazz licks
- cellProgress: collection containing the user's self-rated progress for each cell (lick). one entry for each training session
- cellCollection: view collection, currently joining cells and cellProgress but not user agnostic

My goal is to offer the user a list with all cells and for the cells that they already have progress on to show them their personal progress (as indicated by Leitner-Box [a system for spaced repetition] and # of repetitions in cell progress).

Now I'm wondering if I need to build this on the frontend by fetching both the user specific progress and all cells or whether there's a way to let pocketBase handle this 🤔 

[pb_schema.json](https://github.com/pocketbase/pocketbase/files/14736142/pb_schema.json)

Thanks a lot for your help, much appreciated!

**Reply by ganigeorgiev** [reply]
I'm not familiar with the music related terminology and it is possible that I'm misunderstanding something but I don't think a View collection is needed here. My suggestions:

- Change the `cellProgress.cellId` field to be a **single nonempty** `relation` field pointing to the `cells` collection (_you'll have to remove your existing `text` field as field type changes are not allowed; the nonempty constraint can be enforced from the field options_).
- Change the `cellProgress.userId` field to be a **single nonempty** `relation` field pointing to the users collection (_you'll have to remove your existing `number` field as field type changes are not allowed; the nonempty constraint can be enforced from the field options_).
- If a single user can have a `cellProgress` to a specific `cells` only once, then define a Unique combined index by selecting both `cellId` and `userId` field from the "Unique constraints and indexes" section under the fields.
- Specify a List and View API rules in the `cellProgress` collection that will allow only the current authenticated user to see their progress. One such possible rule could be:
   ```js
   @request.auth.id = userId
   ```

By doing the above you'll have the option to use [Back-relation expand](https://pocketbase.io/docs/working-with-relations/#back-relations) like this:

```js
await pb.collection("user").authWithPassword("test@example.com", "1234567890")

const result = await pb.collection("cells").getList(1, 20, {
    expand: "cellProgress_via_cellId"
})
```
The above should return something like:
```js
{
  "page": 1,
  "perPage": 20,
  "totalPages": 3,
  "totalItems": 60,
  "items": [
    {
      "id": "RECORD_ID",
      "collectionId": "y98w2sr8iohb0la",
      "collectionName": "cells",
      "created": "2022-01-01 01:00:00.123Z",
      "updated": "2022-01-01 23:59:59.456Z",
      "cellTitle": "test",
      "notes": "JSON",
      "firstNote": "test",
      "lastNote": "test",
      "keyCenters": [ "RELATION_RECORD_ID" ],
      "subtitle": "test"
      // will be empty if the user doesn't have a cellProgress for the specific cell
      "expand": {
        "cellProgress_via_cellId": [
          {
            "id": "RECORD_ID",
            "collectionId": "6mboi4ykg0os6mk",
            "collectionName": "cellProgress",
            "created": "2022-01-01 01:00:00.123Z",
            "updated": "2022-01-01 23:59:59.456Z",
            "userId": "USER_RECORD_ID",
            "cellId": "CELL_RECORD_ID",
            "leitnerBox": 123,
            "repetitions": 456
          }
        ]
      }
    },
    ...
  ]
}
```

**Comment by hrr-amboss** [comment]
@ganigeorgiev Wow, this is amazing. Thanks so much for your guidance. I will try this out and report the final solution I found here.
